### ISSUE TITLE
bug: Coq's autocomplete popup is aligned to the left of the screen

### ISSUE THREAD

**Issue by dapetcu21** [folke/noice.nvim#967]
### Did you check docs and existing issues?

- [X] I have read all the noice.nvim docs
- [X] I have updated the plugin to the latest version before submitting this issue
- [X] I have searched the existing issues of noice.nvim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

0.10.1

### Operating system/version

Windows 10

### Describe the bug

With noice installed, Coq's autocomplete popup menu is always aligned to the left of the screen instead of being aligned to the autocomplete options window.

A workaround is `opts = { popupmenu = { enabled = false} }`, but that disables other useful features like the command palette.

![image](https://github.com/user-attachments/assets/3d78182a-d305-4bb4-9ad4-1be4f50fc658)

### Steps To Reproduce

Have nvim-coq and noice installed, with pretty much default settings, then try to highlight an autocomplete option.

The minimal repro below works, but, for reference, my vim config is at https://github.com/dapetcu21/vimrc
Noice config here: https://github.com/dapetcu21/vimrc/blob/main/lua/userconf/plugins/ui.lua
COQ config here: https://github.com/dapetcu21/vimrc/blob/main/lua/userconf/plugins/lsp.lua

### Expected Behavior

![image](https://github.com/user-attachments/assets/d2d4c008-9a54-4b49-aab6-99ff5bc26b74)

### Repro

```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

require("lazy.minit").repro({
  spec = {
    { "folke/noice.nvim", opts = {} }, -- can be worked around with opts = { popupmenu = { enabled = false } }
    {
      'ms-jpq/coq_nvim',
      branch = 'coq',
      build = ':COQdeps',
      init = function ()
        vim.g.coq_settings = { auto_start = 'shut-up' }
      end,
    },

    {
      'ms-jpq/coq.artifacts',
      branch = 'artifacts',
      dependencies = { 'ms-jpq/coq_nvim' }
    },

  },
})
```

**Comment by GitMurf** [comment]
I have a similar problem where I use the neovim native omnifunc for lsp and have styled it accordingly, but when noice is installed it changes the styling of my completions window and also puts the docs window to the left side of the screen (like above user reported). The problem is if I disable popupmenu in noice config, then the cmdline completions does not use the normal completions popup and instead just shows all the Tab options at the bottom in the statusline row horizontally aligned to Tab through (very hard to review).

Without noice, everything works  as expected including my completions popup in cmdline. I just basically want that functionality (noice not to do anything with popupmenu / native omnifunc window). Is this possible?

**Comment by GitMurf** [comment]
To detail what I mean, see the following:

Completion list/menu in cmdline with noice installed and popupmenu config disabled

![image](https://github.com/user-attachments/assets/5ab7c886-e551-4be0-b617-f24fab605c51)

Here is what my cmdline completions look like with noice DISABLED. This is what I want. Just keep my "old" functionality.

![image](https://github.com/user-attachments/assets/40869993-ccd7-42fa-beff-4f5d6ad46c3e)

**Comment by folke** [comment]
how is this a Noice issue?
Noice has no custom integration for coq (and will never have).
Coq just assumes your cmdline is still at its notmal place.
Closing as won't fix.

**Comment by dapetcu21** [comment]
@folke wait! This is not about the cmdline, but the position of the docs window relative to the autocomplete window. And also this is apparently not even about coq specifically. As the previous user mentioned, it happens when using the native omnifunc, so I think it's worth at least an investigation before being dismissed.

**Comment by folke** [comment]
Native omnifunc with native neovim popup-menu works as it should. That doc window is coq, is it not?

**Comment by folke** [comment]
I don't even use CMP and native neovim popup menus all work as they should.

**Comment by GitMurf** [comment]
@folke I agree with @dapetcu21 that this is not about coq specifically.

Whatever Noice is doing when popupmenu is enabled in noice config, it makes changes to the default omnifunc popup for regular completions in regular buffers. Additionally as @dapetcu21 mentioned, when popupmenu is enabled it also messes with the positioning of the docs window for my regular lsp omnifunc completions. See below:

Here is how a regular completion looks for me with my neovim native lsp omnifunc (and the setup / styling I have for it):

- Notice the highlighting colors etc.
- Notice the highlighted matched characters (red in first screenshot, pink in second)
- Notice the window positioning for docs
- Notice the ordering within each completion menu item (kind, deprecation and lsp source)

![image](https://github.com/user-attachments/assets/3cd721ed-05b9-4f56-81c3-ac1b4e080eef)

Now here is what happens when I simply enable popupmenu for cmdline in noice (the default simply doing `popupmenu = { enabled = true }`. Notice the 3 items identified above and how they are now different.

![image](https://github.com/user-attachments/assets/e26998b7-a987-4317-bdc9-0b17a36aec86)

Let me know if you have any questions and need any further clarity.

**Comment by folke** [comment]
Sorry for my ignorance, but how do you get those doc windows natively with omnifunc?
Can you give me some code I can test so I can look at this?
`ui_attach` doesn't yield messages related to doc windows, so I have no control over positioning those I think. No idea tbh. Need to be able to test it first.

**Comment by GitMurf** [comment]
For the record, I do not really care about the highlighting issues etc. if it only applies to the completions menu in the cmdline. The completions I get in the cmdline with popupmenu enabled are fine for me as is, the problem is I don't want any of the popupmenu stuff affecting my completions menu outside of the cmdline. Does that make sense?

**Comment by folke** [comment]
You're both talking about that doc window being in the wrong position.
Where does that window come from and how can I test that?

**Comment by GitMurf** [comment]
> You're both talking about that doc window being in the wrong position. Where does that window come from and how can I test that?

I did a lot of my own configuration of some of the newer neovim completion primitives available so instead of trying to give you that, the same issue arises when using mini.completion out of the box. Should be much easier to quickly test and then review his code since it is encapsulated in a single file.

I just did a simple: `require('mini.completion').setup({})` to reproduce the screenshot below (with noice popupmenu disabled). And then if you enable popupmenu with noice the docs window is shifted all the way to the left (as in our earlier comments).

![image](https://github.com/user-attachments/assets/f3b16192-d1db-45e3-8cd5-941c14f158e3)

**Comment by folke** [comment]
Are you referring to signature help? So not hover docs?

**Comment by GitMurf** [comment]
> Are you referring to signature help? So not hover docs?

No, these are lsp docs. as I cycle through each completion item I get the docs for each function / method / variable etc. coming from LSP. It looks a bit weird in the screenshot because it is updating my text each time I cycle to a new item, but as you can see from the highlighted red text in completion menu, I had only types `nvim_get_o` and then <c-n> down to the item.

**Comment by folke** [comment]
ok, but again. omnifunc / popupmenu, don't have functionality to show docs afaik? So who is creating that popup then?

Or more clearly, who is **positioning** that doc window?

**Comment by folke** [comment]
Allright, I further looked into this and it's clearly a Neovim bug.

`mini.completion` and probably `coq` listen on `CompleteChanged` and then use `vim.v.event.col` and others to determine the position of the pummenu.

It appears that with `ui_attach`, `col` is always `zero`.

There's no way for me to update `vim.v.event`, since that is readonly.

Should be reported and fixed in core. Like I thought, this is not caused by noice.

**Comment by GitMurf** [comment]
Ok thanks for the follow-up as I was going to go try to create a minimal setup for my lsp completions and docs which would take a little time :)

I think I understand what you are saying and that is fine for the docs window alignment issue.

But can you see the other items I mentioned and showed in my screenshot comparison? the highlighting, completion menu format etc.? Those are being changed for my regular omnifunc completions NOT in the cmdline.

TLDR: the docs window and the completions menu formatting would all be solved if you could only attach your popupmenu changes to completions coming inside the cmdline. Is this possible to do? Since the cmdline is just a window, are you able to either A) attach "locally" based on the window / buffer of the cmdline popup window?
or B) able to attach / update when cmdline is shown and then detach the changes when leaving cmdline?

I guess I should first clarify if what I am explaining / asking makes sense or if I need to do a better job explaining? Thanks!

**Comment by folke** [comment]
Noice has it's own implementation of the popupmenu that you can format / style. It doesn't (can't) just use the normal styling.

> Is this possible to do?
Nope, not possible

`ui_attach` is all or nothing.

**Comment by folke** [comment]
As for those multiple fuzzy matches in the same item, I only don't get that info through the API.

**Comment by GitMurf** [comment]
> > Nope, not possible
> `ui_attach` is all or nothing.

Ok then I guess I'll backup to this initial screenshot / question to simplify it all: https://github.com/folke/noice.nvim/issues/967#issuecomment-2465210681

Do you know why when I have Noice plugin installed and disable popupmenu, that the Tab completions come in in a single row at the bottom like my screenshot? As opposed to my second screenshot which shows a proper "regular" completions popup (native neovim) when I disable Noice all together (not just disable popupmenu config)?

Basically it seems like even when disabling popupmenu in Noice config, there are still some settings / changes that are preventing the normal native tab completions menu?

**Comment by folke** [comment]
If you disable popupmenu, then Neovim will use its internal stuff, but won't know the cmdline is actually shown in another place. Similar issue to that of completions.
But it used to be that you could not disable popupmenu independently from cmdline. Maybe that changed

**Comment by GitMurf** [comment]
Right now there is a choice for nui vs cmp for the popupmenu. I wonder if what we need here is essentially a third source similar to cmp which is for using "native" neovim completion menu?

**Comment by folke** [comment]
That's is what noice is. Even when you choose `cmp` as the backend, what happens is that Noice builds the popupmenu using some classes from cmp. Cmp itself never processes anything from that popupmenu.

**Comment by folke** [comment]
You probably don't fully understand what `ui_attach` is.

With `ui_attach`, Noice **IS** the **native** cmdline, the popupmenu, messages, ....

**Comment by GitMurf** [comment]
> You probably don't fully understand what `ui_attach` is.
> 
> With `ui_attach`, Noice **IS** the **native** cmdline, the popupmenu, messages, ....

I definitely do not know much about `ui_attach`. But I more or less assumed that generally... and that is kind of what I was getting at earlier. Can we make sure when disabling popupmenu, that it is completely disabled and therefore to your point above, Neovim would be the `native` popupmenu still / again.

To that point it seems like maybe the complication is between using Noice's `cmdline` but just using Neovim "core" (avoiding usage of the word native) for the popupmenu. Is this true? is this the challenge? Disambiguating completion/popupmenu from the cmdline...

**Comment by folke** [comment]
This is gonna be my last message here. Check the neovim docs on ui attach, specifically the UI attach events. That's the only thing I have to work with. If you disable noice popup-menu, then neovim will use its own, but it won't know where it should render it. Like I said a gazillion times already, nothing I can do about that. To make these things work, the UI attach API needs to be expanded. 

I can't fix this.

**Comment by GitMurf** [comment]
k thanks for your time. sorry for all the comments.
